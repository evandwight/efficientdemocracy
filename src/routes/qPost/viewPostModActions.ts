
import DataFrame from 'dataframe-js';
import validator from 'validator';
import * as C from '../../constant';
import db from '../../db/databaseApi';
import DemocraticModerationService from '../../services/democraticModerationService';
import { ViewPostModAction, VoteDetails } from '../../views/post/viewPostModAction';
import { reactRender } from '../../views/utils';
import { unexpectedAssert, validationAssert } from '../utils';
export * from './listings';

export async function viewPostModAction(req, res) {
    const { id, field } = req.params;
    validationAssert(validator.isUUID(id, 4), "Invalid post id", 400);
    validationAssert(C.FIELDS.LIST.indexOf(field) !== -1, "Invalid field", 400);

    const post = await db.qPosts.getPost(id);
    validationAssert(!!post, "Post not found", 404);

    const action = await DemocraticModerationService.getModAction({ thingId: id, field });
    const { priority } = action || {};
    const { MINI_MOD, SAMPLE_1 } = C.MOD_ACTIONS.PRIORTY;
    let details = null;
    let includeChartJs = false;
    let dangerousChartData;
    if (priority === MINI_MOD || priority >= SAMPLE_1) {
        let isSample, counts, dataTable;
        if (priority == MINI_MOD) {
            counts = await DemocraticModerationService.getMiniModVoteCount({ thingId: id, field });
            isSample = false;
        } else {
            const sample = await DemocraticModerationService.getSampleResult(action.creator_id);
            unexpectedAssert(!!sample, "Sample not found");
            unexpectedAssert(sample.is_complete, "Sample must be complete to view");
            ({ counts } = sample);
            isSample = true;
        }

        ({ dangerousChartData, dataTable } = countsToViewData({ counts, result: action.value, isSample }));
        details = VoteDetails({ dataTable });
        includeChartJs = true;
    }

    reactRender(res, ViewPostModAction({ post, action, field, details }), 
        { title: `Mod info - ${field}`, includeVotesJs: true, includeChartJs, dangerousChartData });
}

export function countsToViewData({ counts, result, isSample }) {
    // Do not put user input into this data. This gets injected into a script tag
    // Be careful what strings you use, I have not escaped this so for example </script> would break the page

    const rows = [true, false, ...isSample ? [null] : []];
    const rowLabel = ["True", "False", ...isSample ? ["Didn't vote"] : []];

    const STRIKE_COLUMNS = ["strike_ups", "strike_downs", "strike_poster", ...isSample ? ["strike_disputers"] : []];

    let df = new DataFrame(counts, ["vote", ...STRIKE_COLUMNS, "count"]);
    const strikeDf = df.filter({ vote: result });

    const dangerousChartData: any = {
        voteCanvas: {
            labels: rowLabel,
            datasets: [{
                label: '# of Votes',
                data: rows.map(v => df.filter({ "vote": v }).stat.sum("count")),
            }]
        },
        strikeCanvas: {
            labels: ["Up voters", "Down voters", "Poster", ...isSample ? ["Disputers"] : []],
            datasets: [{
                label: '# of Strike Votes',
                data: STRIKE_COLUMNS.map(col => strikeDf.filter({ [col]: true }).stat.sum("count")),
            }]
        }
    };
    
    const dataTable = [
        ["Set field to", "Total", "Strike up voters", "Strike down voters", "Strike poster", ...isSample ? ["Strike disputers"] : []],
        ...rows.map((row, i) => [
            rowLabel[i],
            df.filter({ vote: row }).stat.sum("count"), // Total
            ...STRIKE_COLUMNS.map(col => df.filter({ vote: row, [col]: true }).stat.sum("count"))
        ])
    ];
    return { dangerousChartData, dataTable };
}