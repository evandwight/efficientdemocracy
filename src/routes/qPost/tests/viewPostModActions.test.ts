import {countsToViewData} from '../viewPostModActions';

describe('viewPostModActions', () => {
    describe('countsToViewData', () => {
        it('includes disputers', () => {
            const counts = [
                {vote: true, strike_ups: false, strike_downs: true, strike_poster: false, strike_disputers: false, count: 1},
                {vote: false, strike_ups: true, strike_downs: false, strike_poster: true, strike_disputers: true, count: 2},
                {vote: null, strike_ups: true, strike_downs: true, strike_poster: true, strike_disputers: true, count: 3},
            ]
            const res = countsToViewData({counts, result: false, isSample:true});
            expect(res).toMatchSnapshot();
        });
        it('excludes disputers', () => {
            const counts = [
                {vote: true, strike_ups: true, strike_downs: false, strike_poster: false, count: 1},
                {vote: false, strike_ups: true, strike_downs: true, strike_poster: true, count: 2},
                {vote: null, strike_ups: true, strike_downs: true, strike_poster: true, count: 3},
            ]
            const res = countsToViewData({counts, result: true, isSample:false});
            expect(res).toMatchSnapshot();
        });
        it('handles no counts', () => {
            const counts = [];
            const res = countsToViewData({counts, result: false, isSample:true});
            expect(res).toMatchSnapshot();
        });
    });
});