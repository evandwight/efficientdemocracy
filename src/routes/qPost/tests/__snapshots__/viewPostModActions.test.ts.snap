// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`viewPostModActions countsToViewData excludes disputers 1`] = `
Object {
  "dangerousChartData": Object {
    "strikeCanvas": Object {
      "datasets": Array [
        Object {
          "data": Array [
            1,
            0,
            0,
          ],
          "label": "# of Strike Votes",
        },
      ],
      "labels": Array [
        "Up voters",
        "Down voters",
        "Poster",
      ],
    },
    "voteCanvas": Object {
      "datasets": Array [
        Object {
          "data": Array [
            1,
            2,
          ],
          "label": "# of Votes",
        },
      ],
      "labels": Array [
        "True",
        "False",
      ],
    },
  },
  "dataTable": Array [
    Array [
      "Set field to",
      "Total",
      "Strike up voters",
      "Strike down voters",
      "Strike poster",
    ],
    Array [
      "True",
      1,
      1,
      0,
      0,
    ],
    Array [
      "False",
      2,
      2,
      2,
      2,
    ],
  ],
}
`;

exports[`viewPostModActions countsToViewData handles no counts 1`] = `
Object {
  "dangerousChartData": Object {
    "strikeCanvas": Object {
      "datasets": Array [
        Object {
          "data": Array [
            0,
            0,
            0,
            0,
          ],
          "label": "# of Strike Votes",
        },
      ],
      "labels": Array [
        "Up voters",
        "Down voters",
        "Poster",
        "Disputers",
      ],
    },
    "voteCanvas": Object {
      "datasets": Array [
        Object {
          "data": Array [
            0,
            0,
            0,
          ],
          "label": "# of Votes",
        },
      ],
      "labels": Array [
        "True",
        "False",
        "Didn't vote",
      ],
    },
  },
  "dataTable": Array [
    Array [
      "Set field to",
      "Total",
      "Strike up voters",
      "Strike down voters",
      "Strike poster",
      "Strike disputers",
    ],
    Array [
      "True",
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      "False",
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      "Didn't vote",
      0,
      0,
      0,
      0,
      0,
    ],
  ],
}
`;

exports[`viewPostModActions countsToViewData includes disputers 1`] = `
Object {
  "dangerousChartData": Object {
    "strikeCanvas": Object {
      "datasets": Array [
        Object {
          "data": Array [
            2,
            0,
            2,
            2,
          ],
          "label": "# of Strike Votes",
        },
      ],
      "labels": Array [
        "Up voters",
        "Down voters",
        "Poster",
        "Disputers",
      ],
    },
    "voteCanvas": Object {
      "datasets": Array [
        Object {
          "data": Array [
            1,
            2,
            3,
          ],
          "label": "# of Votes",
        },
      ],
      "labels": Array [
        "True",
        "False",
        "Didn't vote",
      ],
    },
  },
  "dataTable": Array [
    Array [
      "Set field to",
      "Total",
      "Strike up voters",
      "Strike down voters",
      "Strike poster",
      "Strike disputers",
    ],
    Array [
      "True",
      1,
      0,
      1,
      0,
      0,
    ],
    Array [
      "False",
      2,
      2,
      0,
      2,
      2,
    ],
    Array [
      "Didn't vote",
      3,
      3,
      3,
      3,
      3,
    ],
  ],
}
`;
